hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java

```diff
@@ -1245,6 +1253,7 @@ void scheduleBlockReportInt(long delay) {
       lastBlockReport = 0;
     }
     resetBlockReportTime = true; // reset future BRs for randomness
+    nextBlockReportOverwritten = true; //make sure that if there is an ongoing blockreport it will not cancel this
```

Comment:
Everything seems fine except that I am unable to understand how this prevents cancelling of a ongoing block report. if a blockreport is happening then the next one will be put on hold because we are using Executors.newSingleThreadExecutor() for sending block reports

Reply:
Yes, the new block report should be on hold and start immediately after the current one finish. The problem is that in the current implementation when the current block report finishes it overwrite the time at which the next block report should happen and as a result, the block report on hold is not triggered immediately after. This new condition is here to stop the ongoing block report from overwriting the scheduled time for the next block report.